# **Ananya Gupta**

**Dataset Used** : Records Matching Task - data1.csv & 
                 Records Matching Task - data2.csv
 **https://docs.google.com/spreadsheets/d/1g99QKwm2V-_JMXtup7Nk081axpxRYbBx/edit?usp=sharing&ouid=103676608938964511475&rtpof=true&sd=true**

In [9]:
# Importing the required libraries
import numpy as np
import pandas as pd

**Reading data1 table from the given dataset to Pandas Dataframe**

In [4]:
data1 = pd.read_csv('Records Matching Task - data1.csv')

**Combining Order Id + Product ID into one unique column**

In [5]:
data1['Unique ID1'] = data1['Order ID'] + " " + data1['Product ID']

In [6]:
# Printing the first 5 rows of data1
data1.head()

Order ID       Product ID  Qty                      Unique ID1
0  CA-2014-100006  TEC-PH-10002075    3  CA-2014-100006 TEC-PH-10002075
1  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090 FUR-TA-10003715
2  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090 OFF-BI-10001597
3  CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293 OFF-PA-10000176
4  CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328 OFF-BI-10000343

**It’s imperative to know the count of NaN values for each column before we proceed further.**

In [8]:
data1.isnull().sum()

Order ID      0
Product ID    0
Qty           0
Unique ID1    0
dtype: int64

**Similarly, for data2 in the given dataset combine the Order ID + Product ID into one column to make a Unique ID for data2**

In [11]:
data2 = pd.read_csv('Records Matching Task - data2.csv')

In [12]:
data2['Unique ID2'] = data2['Order ID'] + " " + data2['Product ID']

In [14]:
data2.head()

Order ID       Product ID  Qty                      Unique ID2
0  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090 FUR-TA-10003715
1  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090 OFF-BI-10001597
2  CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293 OFF-PA-10000176
3  CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328 OFF-BI-10000343
4  CA-2014-100363  OFF-FA-10000611    2  CA-2014-100363 OFF-FA-10000611

In [15]:
# check if there exists any null values in data2
data2.isnull().sum()

Order ID      0
Product ID    0
Qty           0
Unique ID2    0
dtype: int64

**Combining the 2 datasets into one**

In [39]:
combined_data = data1.merge(data2, how = 'outer', left_on = ['Unique ID1'], right_on = ['Unique ID2'])
combined_data.head()

Order ID_x     Product ID_x  Qty_x                      Unique ID1  \
0  CA-2014-100006  TEC-PH-10002075    3.0  CA-2014-100006 TEC-PH-10002075   
1  CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090 FUR-TA-10003715   
2  CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090 OFF-BI-10001597   
3  CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293 OFF-PA-10000176   
4  CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328 OFF-BI-10000343   

       Order ID_y     Product ID_y  Qty_y                      Unique ID2  
0             NaN              NaN    NaN                             NaN  
1  CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090 FUR-TA-10003715  
2  CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090 OFF-BI-10001597  
3  CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293 OFF-PA-10000176  
4  CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328 OFF-BI-10000343

***Question-1***
###***How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)*** 

In [20]:
# Its similar to finding A-B in sets, we just have to find the difference
ans1 = set(data1['Unique ID1']).difference(set(data2['Unique ID2']))
print(len(ans1))

507


**Or you can create an eniterly new dataframe where records present in data1 are missing in data2**

In [42]:
present_in_data1_missing_in_data2 = combined_data[combined_data['Unique ID2'].isna()]
present_in_data1_missing_in_data2.shape[0]

507

***Answer-1***
###***There are total 507 records that are present in data1 but missing in data2.*** 

***Question-2***
###***How to identify the Records (Order ID + Product ID combination) misisng in data1 but present in data2 (Specify the number of records missing in your answer)*** 

In [23]:
# Its similar to finding B-A in sets, we just have to find the difference
ans2 = set(data2['Unique ID2']).difference(set(data1['Unique ID1']))
print(len(ans2))

508


In [44]:
# or you can use method 2
missing_in_data1_present_in_data2 = combined_data[combined_data['Unique ID1'].isna()]
missing_in_data1_present_in_data2.shape[0]

508

***Answer-2***
###***There are total 508 records that are present in data2 but missing in data1.*** 

***Question-3***
###***Find the Sum of the total Qty of Records missing in data1 but present in data2.*** 

In [46]:
sum_of_qty_records = missing_in_data1_present_in_data2['Qty_y'].sum()
print(sum_of_qty_records)

1956.0


***Answer-3***
###***The sum of the total Qty of Records missing in data1 but present in data2 are 1956.*** 

***Question-3***
###***Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2*** 

In [48]:
combined_data.shape[0]

9986

***Answer-4***
###***The total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2 are 9986.*** 